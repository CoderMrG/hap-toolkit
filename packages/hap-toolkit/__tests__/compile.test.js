/*
 * Copyright (C) 2017, hapjs.org. All rights reserved.
 */
'use strict'

const path = require('path')
const del = require('del')
const glob = require('glob')
const { copyApp } = require('hap-dev-utils')
const { compile, stopWatch } = require('../lib')

describe('测试compile', () => {
  const platform = 'native'
  let projectRoot
  let buildDir

  beforeAll(async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/app')
    projectRoot = await copyApp(testAppDir)
    buildDir = path.resolve(projectRoot, 'build')
  })

  it('generate webpack config', () => {
    const genWebpackConf = require('../gen-webpack-conf')
    const conf = genWebpackConf({
      cwd: projectRoot
    })
    expect(conf.entry).toMatchSnapshot()
  })

  it(
    'compile release',
    async () => {
      const mode = 'prod'
      const expectResult = [
        'app.js',
        'CardDemo/index.js',
        'Demo/index.js',
        'workers/request/index.js'
      ]

      const { stats } = await compile(platform, mode, false, {
        cwd: projectRoot
      })
      expect(stats.hasErrors()).toBe(false)

      const result = glob.sync('**/*.{js,js.map}', {
        cwd: buildDir
      })
      expect(result).toEqual(expect.arrayContaining(expectResult))
      expect(result).toMatchSnapshot()
    },
    5 * 60 * 1000
  )

  it(
    'compile build',
    async () => {
      const mode = 'dev'
      const expectResult = [
        'app.js',
        'app.js.map',
        'CardDemo/index.js',
        'CardDemo/index.js.map',
        'Demo/index.js',
        'Demo/index.js.map',
        'workers/request/index.js',
        'workers/request/index.js.map'
      ]

      // 第三个参数为是否开启watch，true为开启
      const data = await compile(platform, mode, true, { cwd: projectRoot })
      expect(data.compileError).toBeNull()
      expect(data.stats.hasErrors()).toBe(false)

      const result = glob.sync('**/*.{js,js.map}', {
        cwd: buildDir
      })
      expect(result).toEqual(expect.arrayContaining(expectResult))
      expect(result).toMatchSnapshot()
    },
    5 * 60 * 1000
  )

  it('stop watch', async () => {
    const data = await stopWatch()
    expect(data.stopWatchError).toBeNull()
  })

  afterAll(async () => {
    await del([projectRoot])
  })
})
