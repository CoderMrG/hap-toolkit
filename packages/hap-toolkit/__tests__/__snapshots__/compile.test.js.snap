// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试compile compile build 1`] = `
Array [
  "app.js",
  "app.js.map",
  "CardDemo/index.js",
  "CardDemo/index.js.map",
  "Demo/index.js",
  "Demo/index.js.map",
  "workers/request/index.js",
  "workers/request/index.js.map",
]
`;

exports[`测试compile compile release 1`] = `
Array [
  "app.js",
  "CardDemo/index.js",
  "Demo/index.js",
  "workers/request/index.js",
]
`;

exports[`测试compile generate webpack config 1`] = `
Object {
  "CardDemo/index": "./src/CardDemo/index.ux?uxType=card",
  "Demo/index": "./src/Demo/index.ux?uxType=page",
  "app": "./src/app.ux?uxType=app",
  "workers/request/index": "./src/workers/request/index.js",
}
`;
